### DISCUSSION TITLE
Struggling with rules precedence

### DISCUSSION THREAD

**Discussion by biozz** [tree-sitter/tree-sitter#1721] [Q&A] [answered]
Hi!

I am trying to implement tree-sitter for [todo.txt](http://todotxt.org/) file format and struggling with rules precedence. I feel like I am missing something and hope the community can help.

This test case fails:

```
=================================================
Complited task
=================================================
x task

-------------------------------------------------
(todotxt
  (task
    (done)))
```

My grammar definition doesn't allow tree-sitter to recognize `x something` (completed task). The actual result is:

```
(todotxt
  (task))
```

This is what I've tried so far:

- setting various `prec` directives, but there is no actual conflict and they don't work
- moving the rules around and sometimes it hangs in infinite loops
- adding negative lookups to word symbols, but that complicates things a lot

Any ideas?

Links:

- [todo.txt file format](https://github.com/todotxt/todo.txt)
- [biozz/tree-sitter-todotxt grammar.js](https://github.com/biozz/tree-sitter-todotxt/blob/main/grammar.js)

**Comment by ahelwer** [comment, marked as answer]
Lexical precedence. Specify 'prec' inside the 'token' function.

**Reply by biozz** [reply]
That worked! Thank you.

Do you mind if I squeeze in another question?

This test:

```
=================================================
A task with project
=================================================
task +project hello

-------------------------------------------------
(todotxt
  (task
    (project)))
```

results in:

```
(todotxt
  (task))
```

It looks like the `_word` eaten the project and `/\+\w+/` was never captured.

Do you know if this is also related to properly placed prec directive?

**Reply by ahelwer** [reply]
It helps to look at the parse debug output to get an idea of what your parser is doing. It might be a case of a regex greedily eating up as much text as it can.

**Reply by biozz** [reply]
@ahelwer I followed your advice and looked through debug logs. Indeed, `_word` pattern with `/.*/` was capturing too much. I changed it to `/\S*/` and it works perfectly now. Thank you again for your support!
